src/components/svgComponent.ts

```diff
@@ -0,0 +1,354 @@
+    } else {
+      this._element = this._rootSVG.append("g");
+      this._setup();
+    }
+
+    this._isAnchored = true;
+    this._onAnchorCallbacks.callCallbacks(this);
+    return this;
+  }
+
+  /**
+   * Creates additional elements as necessary for the Component to function.
+   * Called during anchor() if the Component's element has not been created yet.
+   * Override in subclasses to provide additional functionality.
+   */
+  protected _setup() {
+    if (this._isSetup) {
+      return;
+    }
+    if (this.parent() != null) {
+      // this component is a top level SVG; however, it has parents
+      // which means that it is nested within a div container
+      this._rootSVG.classed("component", true);
+    }
+
+    this._cssClasses.forEach((cssClass: string) => {
+      this._element.classed(cssClass, true);
+    });
+
+    this._cssClasses = new Utils.Set<string>();
+
+    this._backgroundContainer = this._element.append("g").classed("background-container", true);
+    this._addBox("background-fill", this._backgroundContainer);
+    this._content = this._element.append("g").classed("content", true);
+    this._foregroundContainer = this._element.append("g").classed("foreground-container", true);
+    this._boxContainer = this._element.append("g").classed("box-container", true);
+
+    if (this._clipPathEnabled) {
+      this._generateClipPath();
+    }
+
+    this._boundingBox = this._addBox("bounding-box");
+    this._isSetup = true;
+  }
+
+  /**
+   * Computes and sets the size, position, and alignment of the Component from the specified values.
+   * If no parameters are supplied and the Component is a root node,
+   * they are inferred from the size of the Component's element.
+   *
+   * @param {Point} [origin] Origin of the space offered to the Component.
+   * @param {number} [availableWidth] Available width in pixels.
+   * @param {number} [availableHeight] Available height in pixels.
+   * @returns {Component} The calling Component.
+   */
+  public computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number) {
+    if (origin == null || availableWidth == null || availableHeight == null) {
+      if (this._element == null) {
+        throw new Error("anchor() must be called before computeLayout()");
+      } else if (this.parent() == null) {
+        // we are the root node, retrieve height/width from root SVG
+        origin = { x: 0, y: 0 };
+
+        // Set width/height to 100% if not specified, to allow accurate size calculation
+        // see http://www.w3.org/TR/CSS21/visudet.html#block-replaced-width
+        // and http://www.w3.org/TR/CSS21/visudet.html#inline-replaced-height
+        if (this._rootSVG.attr("width") == null) {
+          this._rootSVG.attr("width", "100%");
+        }
+        if (this._rootSVG.attr("height") == null) {
+          this._rootSVG.attr("height", "100%");
+        }
+
+        let elem: HTMLScriptElement = (<HTMLScriptElement> this._rootSVG.node());
```

Comment:
Why is this cast to an `HTMLScriptElement`? Shouldn't this be a `SVGElement`, or actually a `SVGSVGElement`